### PR TITLE
New example showing pynamodb, pre-signed urls, s3 events and log config.

### PR DESCRIPTION
**Pull request by bedge**
This example is a derivation of the aws-python-rest-api-with-pynamodb but it adds many other features and shows alternate ways or arranging the service's serverless.yml to make it more DRY.
It also show a simple way to customize the lambda/python logging to add file:line_num and other details in a single place for all handlers.

### CODE REVIEW

File: aws-python-pynamodb-s3-sigurl/README.md [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+description: This example demonstrates how to setup a RESTful Web Service allowing you to create, list, get, update and delete Assets. DynamoDB is used to store the data.
+layout: Doc
+-->
+# Serverless REST API
+This example demonstrates how to setup a [RESTful Web Service](https://en.wikipedia.org/wiki/Representational_state_transfer#Applied_to_web_services) 
+using [Presigned URLs](http://boto3.readthedocs.io/en/latest/guide/s3.html?highlight=presigned#generating-presigned-urls) 
+to manage asset uploads and downloads. 
+
+The initial POST creates an asset entry in dynamo and returns a presigned upload URL. 
+This is used upload the asset without needing any credentials.
```
Comment (christophgysin) [inline aws-python-pynamodb-s3-sigurl/README.md]: *to

File: aws-python-pynamodb-s3-sigurl/README.md [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+-->
+# Serverless REST API
+This example demonstrates how to setup a [RESTful Web Service](https://en.wikipedia.org/wiki/Representational_state_transfer#Applied_to_web_services) 
+using [Presigned URLs](http://boto3.readthedocs.io/en/latest/guide/s3.html?highlight=presigned#generating-presigned-urls) 
+to manage asset uploads and downloads. 
+
+The initial POST creates an asset entry in dynamo and returns a presigned upload URL. 
+This is used upload the asset without needing any credentials. 
+An s3 event triggers another lambda method to mark the asset as "RECEIVED".
+One can then initiate a PUT to the asset's REST path to mark it as "UPLOADED"
```
Comment (christophgysin) [inline aws-python-pynamodb-s3-sigurl/README.md]: This seems redundant, and sounds like something you would do if you didn't have S3 events that can trigger this.

File: aws-python-pynamodb-s3-sigurl/README.md [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+## Structure
+This service has a separate directory for all the assets operations. 
+For each operation exactly one file exists e.g. `assets/delete.py`. In each of these files there is exactly one function defined.
+### Model
+The idea behind the `assets` directory is that in case you want to create a service containing multiple resources e.g. users, notes, 
+comments you could do so in the same service. 
+While this is certainly possible you might consider creating a separate service for each resource. 
+It depends on the use-case and your preference.
+### API GW Integration model
+The methods that accept no parameters, POST, DELETE, use the default `lambda_proxy` AWS API GW integration.
```
Comment (christophgysin) [inline aws-python-pynamodb-s3-sigurl/README.md]: This does not seems to reflect the code?

File: aws-python-pynamodb-s3-sigurl/README.md [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+```python
+'%(asctime)-15s %(process)d-%(thread)d %(name)s [%(filename)s:%(lineno)d] :%(levelname)8s: %(message)s'
+```
+
+### Notes
+Initial scaffold copied from the aws-python-rest-api-with-pynamodb example.
+
+The PUT method to mark the asset as UPLOADED is somewhat redundant as the S3 event that marks uploads as RECEIVED should be sufficient for most cases.
+However the goal was to use a PUT method to mark it received, so the PUT marks a RECEIVED asset as UPLOADED.
+That said, there is no distinction between UPLOADED vs RECEIVED anywhere in the example.
```
Comment (christophgysin) [inline aws-python-pynamodb-s3-sigurl/README.md]: I think this is slightly confusing. Maybe you could use PUT to provide soft-delete, and have DELETE actually remove the asset?

File: aws-python-pynamodb-s3-sigurl/asset/create.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+                                 'Host': 'c1xblyjsid.execute-api.us-east-1.amazonaws.com', 'User-Agent': 'curl/7.56.1',
+                                 'Via': '1.1 022c901b294fedd7074704d46fce9819.cloudfront.net (CloudFront)',
+                                 'X-Amz-Cf-Id': 'BifKUMLw8qO30TNbJ4QObNGq6WVxiL9nTv9eMbRtAIqqHIqQDkZEVw==',
+                                 'X-Amzn-Trace-Id': 'Root=1-5a1b387c-47ab478111bbb2eb6bd6530c',
+                                 'X-Forwarded-For': '75.82.111.45, 216.137.44.14', 'X-Forwarded-Port': '443',
+                                 'X-Forwarded-Proto': 'https'}, 'query': {}, 'path': {},
+                     'identity': {'cognitoIdentityPoolId': '', 'accountId': '', 'cognitoIdentityId': '', 'caller': '',
+                                  'apiKey': '', 'sourceIp': '75.82.111.45', 'accessKey': '',
+                                  'cognitoAuthenticationType': '', 'cognitoAuthenticationProvider': '', 'userArn': '',
+                                  'userAgent': 'curl/7.56.1', 'user': ''}, 'stageVariables': {}}
```
Comment (christophgysin) [inline aws-python-pynamodb-s3-sigurl/asset/create.py]: Are these just event examples? Could they be comments instead?

File: aws-python-pynamodb-s3-sigurl/asset/get.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+                    'X-Amzn-Trace-Id': 'Root=1-5a1b4655-785e402d33e13e9d533281ef',
+                    'X-Forwarded-For': '75.82.111.45, 216.137.44.103',
+                    'X-Forwarded-Port': '443', 'X-Forwarded-Proto': 'https'},
+        'query': {'timeout': '1000000'},
+        'path': {'asset_id': '0e4e06c6-d2fc-11e7-86c6-6672893a702e'},
+        'identity': {'cognitoIdentityPoolId': '', 'accountId': '', 'cognitoIdentityId': '',
+                     'caller': '', 'apiKey': '', 'sourceIp': '75.82.111.45', 'accessKey': '',
+                     'cognitoAuthenticationType': '', 'cognitoAuthenticationProvider': '',
+                     'userArn': '', 'userAgent': 'curl/7.56.1', 'user': ''},
+        'stageVariables': {}}
```
Comment (christophgysin) [inline aws-python-pynamodb-s3-sigurl/asset/get.py]: same here

File: aws-python-pynamodb-s3-sigurl/serverless.yml [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+          request:
+            paths:
+              asset_id: true
+              parameters:
+                 paths:
+                   timeout: true
+  update:
+    handler: asset/update.update
+    name: ${self:custom.stack_name}-update
+    description: TODO Remove
```
Comment (christophgysin) [inline aws-python-pynamodb-s3-sigurl/serverless.yml]: I guess this sums up my comments above :)

**Comment by bedge** [comment]
Will update as described.

File: aws-python-pynamodb-s3-sigurl/README.md (hunk 2 above)
Comment (bedge) [reply aws-python-pynamodb-s3-sigurl/README.md]: Agreed, it's more to illustrate the point than really being needed. I had a spec I was working form and I took advantage of the s3-events to add some error checking to ensure that one could not 'mark' it (via the PUT) as "UPLOADED" without having actually received a file.
It is a contrived example and as such adds some unneeded complexity to illustrate points.

File: aws-python-pynamodb-s3-sigurl/README.md (hunk 3 above)
Comment (bedge) [reply aws-python-pynamodb-s3-sigurl/README.md]: Good catch, I elected to be consistent on all endpoint and use the same integration for all. Will update docs & resubmit.

File: aws-python-pynamodb-s3-sigurl/README.md (hunk 4 above)
Comment (bedge) [reply aws-python-pynamodb-s3-sigurl/README.md]: The intended goal was to not allow an actual asset deletion from the REST API, however if an authenticated user deleted the s3 asset, then I have no choice but to delete the dynamo entry as well.

File: aws-python-pynamodb-s3-sigurl/asset/create.py (hunk 5 above)
Comment (bedge) [reply aws-python-pynamodb-s3-sigurl/asset/create.py]: Right, yes, I was eval'ing these to validate my indexing at the time. Will correct.

File: aws-python-pynamodb-s3-sigurl/asset/get.py (hunk 6 above)
Comment (bedge) [reply aws-python-pynamodb-s3-sigurl/asset/get.py]: same

File: aws-python-pynamodb-s3-sigurl/serverless.yml (hunk 7 above)
Comment (bedge) [reply aws-python-pynamodb-s3-sigurl/serverless.yml]: Ah, yes, the 'intelligent lint', thanks!

**Comment by bedge** [comment]
@christophgysin This work for you?

**Comment by christophgysin** [comment]
Looks good, thanks for your contribution!

**Pull request merged**